terragrunt 0.67.2 claims terraform AWS S3 state bucket created with previous version not encrypted although SSE-KMS is enabled

## Describe the bug

With terragrunt 0.67.2, running `terragrunt init` in a project with AWS S3 backend created by a previous terragrunt version with encryption enbaled (AWS console shows SSE-KMS active with a KMS key ARN, created with tg 0.67.1), the following prompt appears

```
23:28:12.935 WARN   The remote state S3 bucket xxx needs to be updated:
23:28:12.935 WARN     - Bucket Server-Side Encryption
Remote state S3 bucket xxx is out of date. Would you like Terragrunt to update it? (y/n)
```

When I enter 'y', the following appears

```
23:28:20.777 WARN   Encryption is not enabled on the S3 remote state bucket xxx. Terraform state files may contain secrets, so we STRONGLY recommend enabling encryption!
23:28:20.777 INFO   Enabling Server-Side Encryption for the remote state AWS S3 bucket xxx.
23:28:21.328 INFO   Successfully enabled Server-Side Encryption for the remote state AWS S3 bucket xxx.
```

When I check the config in the AWS console, nothing has changed.

While this could be tolerated for manual interactions, it breaks automated workflows. In our case, GitHub Actions would just stop with unknown exit code:
```
Remote state S3 bucket inventory-quality-gateway-terraform-***-eu-central-1 is out of date. Would you like Terragrunt to update it? (y/n) 21:17:15.507 ERROR  EOF
21:17:15.507 ERROR  Unable to determine underlying exit code, so Terragrunt will exit with error code 1
Error: Process completed with exit code 1.
```

## Steps To Reproduce

Create the following files in your working dir:

terragrunt.hcl:
```hcl
remote_state {
  backend  = "s3"
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }
  config = {
    bucket              = "test-tg-${get_aws_account_id()}"
    key                 = "terraform.tfstate"
    region              = local.aws_region
    dynamodb_table      = "test-tg-${get_aws_account_id()}-lock"
    encrypt             = true
  }
}
```

main.tf:
```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 4.0.0"
    }
  }
}

provider "aws" {
  region = "eu-central-1"
}
```

Then
- Install terraform 1.9.5 and terragrunt 0.67.1
- In your working dir, run `terragrunt init --terragrunt-non-interactive`
- In your working dir, run `terraform apply`

Then upgrade to terragrunt 0.67.2, delete the ./.terraform directory (but keep the generated backend.tf file!) and run `terragrunt init` again in your working dir.

## Expected behavior

No prompt for updating the state bucket / switching on server side encryption should appear when SSE is already enabled.

I would not want to add the --terragrunt-non-interactive flag to the automated execution.

## Versions

- Terragrunt version: 0.67.2
- OpenTofu/Terraform version: terraform 1.9.5
- Environment details: MacOSX Sonoma 14.5, Ubuntu 22.04 (GitHub Actions runner 2.319.1)

--- a/remote/remote_state_s3.go
+++ b/remote/remote_state_s3.go
@@ -332,16 +332,13 @@ func ConfigValuesEqual(config map[string]interface{}, existingBackend *Terraform
 // buildInitializerCacheKey returns a unique key for the given S3 config that can be used to cache the initialization
 func (s3Initializer S3Initializer) buildInitializerCacheKey(
 	s3Config *RemoteStateConfigS3,
-	s3ConfigExtended *ExtendedRemoteStateConfigS3,
 ) string {
 	return fmt.Sprintf(
-		"%s-%s-%s-%s-%s-%s",
+		"%s-%s-%s-%s",
 		s3Config.Bucket,
 		s3Config.Region,
 		s3Config.LockTable,
 		s3Config.DynamoDBTable,
-		s3ConfigExtended.BucketSSEAlgorithm,
-		s3ConfigExtended.BucketSSEKMSKeyID,
 	)
 }
 
@@ -359,7 +356,7 @@ func (s3Initializer S3Initializer) Initialize(ctx context.Context, remoteState *
 
 	var s3Config = s3ConfigExtended.RemoteStateConfigS3
 
-	cacheKey := s3Initializer.buildInitializerCacheKey(&s3Config, s3ConfigExtended)
+	cacheKey := s3Initializer.buildInitializerCacheKey(&s3Config)
 	if initialized, hit := initializedRemoteStateCache.Get(ctx, cacheKey); initialized && hit {
 		terragruntOptions.Logger.Debugf("S3 bucket %s has already been confirmed to be initialized, skipping initialization checks", s3Config.Bucket)
 		return nil
@@ -1361,17 +1358,11 @@ func checkIfSSEForS3MatchesConfig(
 
 	for _, rule := range output.ServerSideEncryptionConfiguration.Rules {
 		if rule.ApplyServerSideEncryptionByDefault != nil && rule.ApplyServerSideEncryptionByDefault.SSEAlgorithm != nil {
-			if *rule.ApplyServerSideEncryptionByDefault.SSEAlgorithm != fetchEncryptionAlgorithm(config) {
-				return false, nil
+			if *rule.ApplyServerSideEncryptionByDefault.SSEAlgorithm == fetchEncryptionAlgorithm(config) {
+				return true, nil
 			}
 
-			if *rule.ApplyServerSideEncryptionByDefault.SSEAlgorithm == s3.ServerSideEncryptionAwsKms &&
-				rule.ApplyServerSideEncryptionByDefault.KMSMasterKeyID != nil &&
-				*rule.ApplyServerSideEncryptionByDefault.KMSMasterKeyID != config.BucketSSEKMSKeyID {
-				return false, nil
-			}
-
-			return true, nil
+			return false, nil
 		}
 	}
 

--- a/test/integration_s3_encryption_test.go
+++ b/test/integration_s3_encryption_test.go
@@ -101,7 +101,15 @@ func TestAwsS3SSECustomKey(t *testing.T) {
 	sseRule = resp.ServerSideEncryptionConfiguration.Rules[0].ApplyServerSideEncryptionByDefault
 	require.NotNil(t, sseRule)
 	assert.Equal(t, s3.ServerSideEncryptionAwsKms, aws.StringValue(sseRule.SSEAlgorithm))
-	assert.True(t, strings.HasSuffix(aws.StringValue(sseRule.KMSMasterKeyID), "alias/other-dedicated-test-key"))
+
+	// This check is asserting that the following bug still isn't fixed:
+	// https://github.com/gruntwork-io/terragrunt/issues/3364
+	//
+	// There were unanticipated consequences to addressing it that should be resolved before the fix is implemented:
+	// https://github.com/gruntwork-io/terragrunt/issues/3384
+	//
+	// At the very least, it should be documented as a breaking change.
+	assert.False(t, strings.HasSuffix(aws.StringValue(sseRule.KMSMasterKeyID), "alias/other-dedicated-test-key"))
 }
 
 func TestAwsS3SSEKeyNotReverted(t *testing.T) {

